File: geopedia/scripts/countries/peru/generate/admin-quiz-data.ts
/**
 * Generates the administrative data used by Peru's quiz configurations.
 *
 * Inputs:
 *   data/intermediate/countries/peru/admin/departments.geojson
 *   data/intermediate/countries/peru/admin/provinces.geojson
 *   data/intermediate/countries/peru/admin/districts.geojson
 *
 * Output:
 *   src/quiz/quizzes/countries/south-america/peru/data/admin.ts
 */

import fs from "node:fs";
import path from "node:path";

type GeoJsonFeature = {
  properties: Record<string, unknown>;
};

type GeoJsonFeatureCollection = {
  type: "FeatureCollection";
  features: GeoJsonFeature[];
};

type Region = {
  id: string;
  name: string;
};

type Province = {
  id: string;
  name: string;
  regionId: string;
};

type District = {
  id: string;
  name: string;
  provinceId: string;
  regionId: string;
};

const ROOT = process.cwd();

const INTERMEDIATE_DIRECTORY = path.join(
  ROOT,
  "data",
  "intermediate",
  "countries",
  "peru",
  "admin",
);

const OUTPUT_PATH = path.join(
  ROOT,
  "src",
  "quiz",
  "quizzes",
  "countries",
  "south-america",
  "peru",
  "data",
  "admin.ts",
);

const REGIONS_INPUT = path.join(
  INTERMEDIATE_DIRECTORY,
  "regions.geojson",
);

const PROVINCES_INPUT = path.join(
  INTERMEDIATE_DIRECTORY,
  "provinces.geojson",
);

const DISTRICTS_INPUT = path.join(
  INTERMEDIATE_DIRECTORY,
  "districts.geojson",
);

// ---------------------------------------------------------------------------
// GeoJSON loading
// ---------------------------------------------------------------------------

function readGeoJson(filePath: string): GeoJsonFeatureCollection {
  const raw = fs.readFileSync(filePath, "utf8");

  const data = JSON.parse(raw) as GeoJsonFeatureCollection;

  if (
    data.type !== "FeatureCollection" ||
    !Array.isArray(data.features)
  ) {
    throw new Error(
      `${filePath} is not a valid GeoJSON FeatureCollection.`,
    );
  }

  return data;
}

function requireString(
  properties: Record<string, unknown>,
  key: string,
): string {
  const value = properties[key];

  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`Missing or invalid string property "${key}".`);
  }

  return value.trim();
}

// ---------------------------------------------------------------------------
// Data loading
// ---------------------------------------------------------------------------

function loadRegions(): Region[] {
  const data = readGeoJson(REGIONS_INPUT);

  return data.features.map((feature) => {
    const properties = feature.properties;

    return {
      id: requireString(properties, "region_id"),
      name: requireString(properties, "region"),
    };
  });
}

function loadProvinces(): Province[] {
  const data = readGeoJson(PROVINCES_INPUT);

  return data.features.map((feature) => {
    const properties = feature.properties;

    return {
      id: requireString(properties, "province_id"),
      name: requireString(properties, "province"),
      regionId: requireString(properties, "region_id"),
    };
  });
}

function loadDistricts(): District[] {
  const data = readGeoJson(DISTRICTS_INPUT);

  return data.features.map((feature) => {
    const properties = feature.properties;

    return {
      id: requireString(properties, "district_id"),
      name: requireString(properties, "district"),
      provinceId: requireString(properties, "province_id"),
      regionId: requireString(properties, "region_id"),
    };
  });
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

function sortById<T extends { id: string }>(values: T[]): T[] {
  return [...values].sort((a, b) => a.id.localeCompare(b.id));
}

function validateUniqueIds<T extends { id: string }>(
  label: string,
  values: T[],
): void {
  const ids = new Set<string>();

  for (const value of values) {
    if (ids.has(value.id)) {
      throw new Error(`Duplicate ${label} ID: ${value.id}`);
    }

    ids.add(value.id);
  }
}

function validateHierarchy(
  regions: Region[],
  provinces: Province[],
  districts: District[],
): void {
  const regionIds = new Set(regions.map((region) => region.id));

  const provinceIds = new Set(
    provinces.map((province) => province.id),
  );

  for (const province of provinces) {
    if (!regionIds.has(province.regionId)) {
      throw new Error(
        `Province ${province.id} references unknown region ` +
          `${province.regionId}.`,
      );
    }
  }

  for (const district of districts) {
    if (!regionIds.has(district.regionId)) {
      throw new Error(
        `District ${district.id} references unknown region ` +
          `${district.regionId}.`,
      );
    }

    if (!provinceIds.has(district.provinceId)) {
      throw new Error(
        `District ${district.id} references unknown province ` +
          `${district.provinceId}.`,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// TypeScript formatting
// ---------------------------------------------------------------------------

function formatRegions(regions: Region[]): string {
  const lines = regions.map(
    (region) =>
      `  ${JSON.stringify(region.id)}: ` +
      `${JSON.stringify(region.name)},`,
  );

  return [
    "export const PERU_REGIONS_BY_ID = {",
    ...lines,
    "} as const;",
  ].join("\n");
}

function formatProvinces(provinces: Province[]): string {
  const lines = provinces.map(
    (province) =>
      `  ${JSON.stringify(province.id)}: { ` +
      `name: ${JSON.stringify(province.name)}, ` +
      `regionId: ${JSON.stringify(province.regionId)} ` +
      `},`,
  );

  return [
    "export const PERU_PROVINCES_BY_ID = {",
    ...lines,
    "} as const;",
  ].join("\n");
}

function formatDistricts(districts: District[]): string {
  const lines = districts.map(
    (district) =>
      `  ${JSON.stringify(district.id)}: { ` +
      `name: ${JSON.stringify(district.name)}, ` +
      `provinceId: ${JSON.stringify(district.provinceId)}, ` +
      `regionId: ${JSON.stringify(district.regionId)} ` +
      `},`,
  );

  return [
    "export const PERU_DISTRICTS_BY_ID = {",
    ...lines,
    "} as const;",
  ].join("\n");
}

// ---------------------------------------------------------------------------
// Output
// ---------------------------------------------------------------------------

function buildOutput(
  regions: Region[],
  provinces: Province[],
  districts: District[],
): string {
  return `/**
 * Generated from Peru's processed administrative GeoJSON.
 *
 * Do not edit manually.
 * Regenerate with:
 * npx tsx scripts/countries/peru/generate/admin-quiz-data.ts
 */

${formatRegions(regions)}

${formatProvinces(provinces)}

${formatDistricts(districts)}
`;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

function main(): void {
  console.log("Generating Peru administrative quiz data...\n");

  const regions = sortById(loadRegions());
  const provinces = sortById(loadProvinces());
  const districts = sortById(loadDistricts());

  validateUniqueIds("region", regions);
  validateUniqueIds("province", provinces);
  validateUniqueIds("district", districts);

  validateHierarchy(regions, provinces, districts);

  if (regions.length !== 25) {
    throw new Error(
      `Expected 25 regions but found ${regions.length}.`,
    );
  }

  if (provinces.length !== 196) {
    throw new Error(
      `Expected 196 provinces but found ${provinces.length}.`,
    );
  }

  if (districts.length !== 1873) {
    throw new Error(
      `Expected 1,873 districts but found ${districts.length}.`,
    );
  }

  const output = buildOutput(regions, provinces, districts);

  fs.mkdirSync(path.dirname(OUTPUT_PATH), {
    recursive: true,
  });

  fs.writeFileSync(OUTPUT_PATH, output, "utf8");

  console.log(`Regions:   ${regions.length}`);
  console.log(`Provinces: ${provinces.length}`);
  console.log(`Districts: ${districts.length}`);

  console.log(`\nGenerated: ${OUTPUT_PATH}`);
}

main();

File: geopedia/scripts/countries/brazil/generate/admin-quiz-data.py
"""
Generates the administrative data used by Brazil's quiz configurations.

Sources:
    public/data/countries/brazil/geojson/regions.geojson
    public/data/countries/brazil/geojson/states.geojson
    public/data/countries/brazil/geojson/municipalities.geojson

Output:
    src/quiz/quizzes/countries/south-america/brazil/data/admin.ts

The processed administrative GeoJSON is the canonical source for region,
state, and municipality IDs, names, and hierarchy.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


PROJECT_ROOT = Path(__file__).resolve().parents[4]

REGIONS_SOURCE_PATH = (
    PROJECT_ROOT
    / "public"
    / "data"
    / "countries"
    / "brazil"
    / "geojson"
    / "regions.geojson"
)

STATES_SOURCE_PATH = (
    PROJECT_ROOT
    / "public"
    / "data"
    / "countries"
    / "brazil"
    / "geojson"
    / "states.geojson"
)

MUNICIPALITIES_SOURCE_PATH = (
    PROJECT_ROOT
    / "public"
    / "data"
    / "countries"
    / "brazil"
    / "geojson"
    / "municipalities.geojson"
)

OUTPUT_PATH = (
    PROJECT_ROOT
    / "src"
    / "quiz"
    / "quizzes"
    / "countries"
    / "south-america"
    / "brazil"
    / "data"
    / "admin.ts"
)

EXPECTED_REGION_COUNT = 5
EXPECTED_STATE_COUNT = 27
EXPECTED_MUNICIPALITY_COUNT = 5_573


def quote(value: str) -> str:
    """Serialize a string as a valid TypeScript string literal."""
    return json.dumps(
        value,
        ensure_ascii=False,
    )


def load_features(
    path: Path,
    label: str,
) -> list[dict[str, Any]]:
    """Load and validate one processed GeoJSON FeatureCollection."""
    if not path.exists():
        raise FileNotFoundError(
            f"{label} GeoJSON was not found: {path}"
        )

    with path.open("r", encoding="utf-8") as file:
        data = json.load(file)

    if data.get("type") != "FeatureCollection":
        raise ValueError(
            f"{label} GeoJSON is not a FeatureCollection."
        )

    features = data.get("features")

    if not isinstance(features, list):
        raise ValueError(
            f"{label} GeoJSON is missing its features array."
        )

    return features


def require_string(
    properties: dict[str, Any],
    property_name: str,
    label: str,
) -> str:
    """Read and validate one required string property."""
    value = properties.get(property_name)

    if not isinstance(value, str) or not value.strip():
        raise ValueError(
            f"{label} is missing a valid {property_name}."
        )

    return value.strip()


def build_regions(
    features: list[dict[str, Any]],
) -> list[tuple[str, str]]:
    """Extract Brazil's region IDs and names."""
    if len(features) != EXPECTED_REGION_COUNT:
        raise ValueError(
            f"Expected {EXPECTED_REGION_COUNT} regions, "
            f"found {len(features)}."
        )

    entries: list[tuple[str, str]] = []
    seen_ids: set[str] = set()

    for index, feature in enumerate(features):
        properties = feature.get("properties")

        if not isinstance(properties, dict):
            raise ValueError(
                f"Region feature {index} has invalid properties."
            )

        region_id = require_string(
            properties,
            "id",
            f"Region feature {index}",
        )

        name = require_string(
            properties,
            "name",
            f"Region {region_id}",
        )

        if region_id in seen_ids:
            raise ValueError(
                f"Duplicate region ID: {region_id}"
            )

        seen_ids.add(region_id)

        entries.append(
            (
                region_id,
                name,
            )
        )

    entries.sort(key=lambda entry: entry[0])

    return entries


def build_states(
    features: list[dict[str, Any]],
) -> list[tuple[str, str, str]]:
    """Extract Brazil's state IDs, names, and parent region IDs."""
    if len(features) != EXPECTED_STATE_COUNT:
        raise ValueError(
            f"Expected {EXPECTED_STATE_COUNT} states, "
            f"found {len(features)}."
        )

    entries: list[tuple[str, str, str]] = []
    seen_ids: set[str] = set()

    for index, feature in enumerate(features):
        properties = feature.get("properties")

        if not isinstance(properties, dict):
            raise ValueError(
                f"State feature {index} has invalid properties."
            )

        state_id = require_string(
            properties,
            "id",
            f"State feature {index}",
        )

        name = require_string(
            properties,
            "name",
            f"State {state_id}",
        )

        region_id = require_string(
            properties,
            "region_id",
            f"State {state_id}",
        )

        if state_id in seen_ids:
            raise ValueError(
                f"Duplicate state ID: {state_id}"
            )

        seen_ids.add(state_id)

        entries.append(
            (
                state_id,
                name,
                region_id,
            )
        )

    entries.sort(key=lambda entry: entry[0])

    return entries


def build_municipalities(
    features: list[dict[str, Any]],
) -> list[tuple[str, str, str]]:
    """Extract municipality IDs, names, and parent state IDs."""
    if len(features) != EXPECTED_MUNICIPALITY_COUNT:
        raise ValueError(
            f"Expected {EXPECTED_MUNICIPALITY_COUNT} municipalities, "
            f"found {len(features)}."
        )

    entries: list[tuple[str, str, str]] = []
    seen_ids: set[str] = set()

    for index, feature in enumerate(features):
        properties = feature.get("properties")

        if not isinstance(properties, dict):
            raise ValueError(
                f"Municipality feature {index} has invalid properties."
            )

        municipality_id = require_string(
            properties,
            "id",
            f"Municipality feature {index}",
        )

        name = require_string(
            properties,
            "name",
            f"Municipality {municipality_id}",
        )

        state_id = require_string(
            properties,
            "state_id",
            f"Municipality {municipality_id}",
        )

        if municipality_id in seen_ids:
            raise ValueError(
                f"Duplicate municipality ID: {municipality_id}"
            )

        seen_ids.add(municipality_id)

        entries.append(
            (
                municipality_id,
                name,
                state_id,
            )
        )

    entries.sort(key=lambda entry: entry[0])

    return entries


def validate_hierarchy(
    regions: list[tuple[str, str]],
    states: list[tuple[str, str, str]],
    municipalities: list[tuple[str, str, str]],
) -> dict[str, str]:
    """
    Validate the administrative hierarchy.

    Returns a state-to-region lookup so municipality entries can include both
    their direct state parent and their derived region parent.
    """
    region_ids = {
        region_id
        for region_id, _ in regions
    }

    state_to_region: dict[str, str] = {}

    for state_id, _, region_id in states:
        if region_id not in region_ids:
            raise ValueError(
                f"State {state_id} references unknown region "
                f"{region_id}."
            )

        state_to_region[state_id] = region_id

    for municipality_id, _, state_id in municipalities:
        if state_id not in state_to_region:
            raise ValueError(
                f"Municipality {municipality_id} references unknown "
                f"state {state_id}."
            )

    return state_to_region


def format_regions(
    regions: list[tuple[str, str]],
) -> str:
    """Format Brazil's region dictionary."""
    lines = "\n".join(
        f"  {quote(region_id)}: {quote(name)},"
        for region_id, name in regions
    )

    return f"""export const BRAZIL_REGIONS_BY_ID = {{
{lines}
}} as const;"""


def format_states(
    states: list[tuple[str, str, str]],
) -> str:
    """Format Brazil's state dictionary."""
    lines = "\n".join(
        (
            f"  {quote(state_id)}: {{ "
            f"name: {quote(name)}, "
            f"regionId: {quote(region_id)} "
            f"}},"
        )
        for state_id, name, region_id in states
    )

    return f"""export const BRAZIL_STATES_BY_ID = {{
{lines}
}} as const;"""


def format_municipalities(
    municipalities: list[tuple[str, str, str]],
    state_to_region: dict[str, str],
) -> str:
    """Format Brazil's municipality dictionary."""
    lines = "\n".join(
        (
            f"  {quote(municipality_id)}: {{ "
            f"name: {quote(name)}, "
            f"stateId: {quote(state_id)}, "
            f"regionId: {quote(state_to_region[state_id])} "
            f"}},"
        )
        for municipality_id, name, state_id in municipalities
    )

    return f"""export const BRAZIL_MUNICIPALITIES_BY_ID = {{
{lines}
}} as const;"""


def create_source(
    regions: list[tuple[str, str]],
    states: list[tuple[str, str, str]],
    municipalities: list[tuple[str, str, str]],
    state_to_region: dict[str, str],
) -> str:
    """Create the generated Brazil administrative data module."""
    return f'''/**
 * Generated from Brazil's processed administrative GeoJSON.
 *
 * Do not edit manually.
 * Regenerate with:
 * python scripts/countries/brazil/generate/admin-quiz-data.py
 */

{format_regions(regions)}

{format_states(states)}

{format_municipalities(municipalities, state_to_region)}
'''


def main() -> None:
    """Generate Brazil's administrative quiz data."""
    print("Generating Brazil administrative quiz data...\n")

    region_features = load_features(
        REGIONS_SOURCE_PATH,
        "Region",
    )

    state_features = load_features(
        STATES_SOURCE_PATH,
        "State",
    )

    municipality_features = load_features(
        MUNICIPALITIES_SOURCE_PATH,
        "Municipality",
    )

    regions = build_regions(
        region_features,
    )

    states = build_states(
        state_features,
    )

    municipalities = build_municipalities(
        municipality_features,
    )

    state_to_region = validate_hierarchy(
        regions,
        states,
        municipalities,
    )

    source = create_source(
        regions,
        states,
        municipalities,
        state_to_region,
    )

    OUTPUT_PATH.parent.mkdir(
        parents=True,
        exist_ok=True,
    )

    OUTPUT_PATH.write_text(
        source,
        encoding="utf-8",
    )

    print(f"Regions:        {len(regions):,}")
    print(f"States:         {len(states):,}")
    print(f"Municipalities: {len(municipalities):,}")
    print(f"\nGenerated: {OUTPUT_PATH}")


if __name__ == "__main__":
    main()